core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -223,6 +224,46 @@ class TransactionStateManager(brokerId: Int,
       throw new IllegalStateException(s"Unexpected empty transaction metadata returned while putting $txnMetadata")))
   }
 
+  def listTransactionStates(
+    filterProducerIds: Set[Long],
+    filterStateNames: Set[String]
+  ): Either[Errors, List[ListTransactionsResponseData.TransactionState]] = {
+    inReadLock(stateLock) {
+      if (loadingPartitions.nonEmpty) {
+        Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
+      } else {
+        val filterStates = filterStateNames.flatMap(TransactionState.fromName)
```

Comment:
Should it produce exception if it include nonexistent state?

Reply:
I could be persuaded probably. Suppose we add a new state in the future. If a user tried to query that state on an old broker, definitely no transactions would exist in that state. From that perspective, the implementation would return an accurate result. On the other hand, I can see how that might be misleading. If we decide to return an error code, then I think we'll have to treat the state filter a little more carefully from a versioning perspective. A new state would probably demand a new version. I think we tried to view the states a little bit more loosely in the `ListGroups` API (which this is modeled after) because we considered the state machine more of an internal implementation detail. This, by the way, is why we went with the string representation rather than numeric ids. This became more of a grey area though when the state filter was added to the request...

Reply:
Probably the response should have a field showing the invalid filters. The callers would know which filters are NOT working.

Reply:
It's a reasonable suggestion. Let me give it a try.

Reply:
How are the states going to be documented (e.g. in the Admin client?) It seems odd to me that on the one hand "they're an impl detail", but they're also used in the filter. How could someone write code to list all of the transaction not in a given state if they can't know the universe of states (given new states could be added in the future)?

Reply:
You can see `listConsumerGroups` to see how this was handled in the analogous case for consumer groups.